picky/src/x509/pkcs7/authenticode.rs

```diff
@@ -0,0 +1,1886 @@
+                         .0
+                        .iter()
+                        .any(|kp_oid| kp_oid == &eku_code_signing_oid)
+                    {
+                        return Err(AuthenticodeError::CaCertificateNotYetValid {
+                            not_before: not_before.into(),
+                            now: now.into(),
+                        });
+                    }
+                } else {
+                    return Err(AuthenticodeError::CaCertificateNotYetValid {
+                        not_before: not_before.into(),
+                        now: now.into(),
+                    });
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn verify(&self) -> AuthenticodeResult<()> {
+        if self.inner.borrow().strictness.require_basic_authenticode_validation {
+            self.verify_authenticode_basic()?;
+        }
+
+        if self.inner.borrow().strictness.require_signing_certificate_check {
+            self.verify_signing_certificate()?;
+        }
+
+        #[cfg(feature = "ctl")]
+        if self.inner.borrow().strictness.require_ca_verification_against_ctl {
+            let ca_name = get_ca_name(&self.authenticode_signature.0.certificates()).unwrap();
+
+            match self.verify_ca_certificate_against_ctl(&ca_name) {
+                Ok(()) => {}
+                Err(err) => {
+                    if !self
+                        .inner
+                        .borrow()
+                        .strictness
+                        .exclude_specific_cert_authorities_from_ctl_check
+                        || !self.inner.borrow().excluded_cert_authorities.contains(&ca_name)
+                    {
+                        return Err(err);
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn check_eku_code_signing(certificates: &[Cert]) -> AuthenticodeResult<()> {
+    if certificates
+        .iter()
+        .flat_map(|cert| cert.extensions().iter())
+        .any(|extension| matches!(extension.extn_value(), ExtensionView::ExtendedKeyUsage(_)))
+    {
+        let signing_cert = certificates.first().ok_or(AuthenticodeError::NoCertificates)?;
+        if !signing_cert
+            .extensions()
+            .iter()
+            .any(|extension| match extension.extn_value() {
+                ExtensionView::ExtendedKeyUsage(eku) => eku.contains(oids::kp_code_signing()),
+                _ => false,
+            })
+        {
+            return Err(AuthenticodeError::NoEKUCodeSigning);
+        }
+    }
+
+    Ok(())
+}
+
+fn get_ca_name(certificates: &[Cert]) -> Option<DirectoryName> {
+    if let Some(root) = certificates.iter().find(|cert| cert.ty() == CertType::Root) {
+        Some(root.subject_name())
+    } else {
+        let intermediate_certificates = certificates
+            .iter()
+            .filter(|cert| cert.ty() == CertType::Intermediate)
+            .cloned()
+            .collect::<Vec<Cert>>();
```

Comment:
Useless clone and heap allocation
```suggestion
            .filter(|cert| cert.ty() == CertType::Intermediate);
```